### PR TITLE
Classically Controlled Short Circuit Bug Fix

### PR DESCRIPTION
**Pull request by ScottCarda-MS**
Rearranged the order of steps in the Classically Controlled transformation so that the restructuring of if-statements is performed first. This addresses an issue where all conditional expressions were always evaluated even when they should have been short-circuited (#475)

### CODE REVIEW

**Comment by cesarzc** [comment]
@ScottCarda-MS if I understand correctly, the basics of this change are the following:
`ReshapeConditionals` has to be executed before `LiftConditionBlocks.Apply` or otherwise all conditional would be evaluated instead of being short-circuited.

In order to do that, another syntax tree transformation (`RestructureConditions`) was created to do so. Now, `ReshapeConditionals` is called as part of this new transformation. Does this explanation make sense?

File: src/QsCompiler/Transformations/ClassicallyControlled.cs [hunk 1]
```diff
@@ -18,22 +18,190 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.ClassicallyControlled
...
-    /// 2nd Pass: On the way down the tree, reshape conditional statements to replace Elif's and
-    /// top level OR and AND conditions with equivalent nested if-else statements. One the way back up
-    /// the tree, convert conditional statements into interface calls, where possible.
-    /// This relies on anything having type parameters must be a global callable.
+    /// This transformation works in three passes.
+    /// 1st Pass: Reshape conditional statements to replace Elif's and top level OR and AND conditions
+    /// with equivalent nested if-else statements.
+    /// 2st Pass: Lift the contents of conditional statements into separate operations, where possible.
+    /// 3nd Pass: Convert conditional statements into interface calls, where possible.
+    /// This relies on global callables being the only things that having type parameters.
```
Comment (cesarzc) [inline src/QsCompiler/Transformations/ClassicallyControlled.cs:26]:
```suggestion
    /// This relies on global callables being the only things that have type parameters.
```

**Comment by ScottCarda-MS** [comment]
> @ScottCarda-MS if I understand correctly, the basics of this change are the following:
> `ReshapeConditionals` has to be executed before `LiftConditionBlocks.Apply` or otherwise all conditional would be evaluated instead of being short-circuited.
> 
> In order to do that, another syntax tree transformation (`RestructureConditions`) was created to do so. Now, `ReshapeConditionals` is called as part of this new transformation. Does this explanation make sense?

Yes, all that is correct and makes sense. It may be worth adding that before the changes in this PR the reshaping conditional logic was part of the `ConvertConditions` transformation that happened after the `LiftConditionBlocks` transformation.

File: src/QsCompiler/Transformations/ClassicallyControlled.cs (hunk 1 above)
Comment (ScottCarda-MS) [reply src/QsCompiler/Transformations/ClassicallyControlled.cs]: Thanks for finding this. It's been fixed.

**Pull request merged**